pkg/dynamicconfig/table.go

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamicconfig
+
+import (
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+
+	"github.com/cilium/cilium/pkg/k8s"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+)
+
+const CiliumConfigTableName = "dynamic-configs"
+
+var ConfigTableIndex = statedb.Index[DynamicConfig, string]{
+	Name: "config-name-key",
+	FromObject: func(t DynamicConfig) index.KeySet {
+		return index.NewKeySet(index.String(t.Name))
+	},
+	FromKey: index.String,
+	Unique:  true,
+}
+var EntryKeyIndex = statedb.Index[DynamicConfig, string]{
+	Name: "config-entry-key",
+	FromObject: func(t DynamicConfig) index.KeySet {
+		return index.StringMap(t.Entries)
+	},
+	FromKey: index.String,
+	Unique:  true,
+}
+
+var ConfigMapKeyEntry = EntryKeyIndex.Query
+
+type DynamicConfig struct {
+	Entries map[string]string
+	Name    string
+}
+
+func NewConfigTable(db *statedb.DB) (statedb.RWTable[DynamicConfig], error) {
+	tbl, err := statedb.NewTable(
+		CiliumConfigTableName,
+		ConfigTableIndex,
+		EntryKeyIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return tbl, db.RegisterTable(tbl)
+}
+
+func NewConfigMapReflector(cs k8sClient.Clientset, c config) k8s.ReflectorConfig[DynamicConfig] {
+	return k8s.ReflectorConfig[DynamicConfig]{
+		Transform: func(a any) (DynamicConfig, bool) {
+			obj := a.(*v1.ConfigMap).DeepCopy()
+			ce := DynamicConfig{Entries: obj.Data, Name: c.DynamicConfigMapName}
+			return ce, true
+		},
+		ListerWatcher: utils.ListerWatcherWithModifiers(
+			utils.ListerWatcherFromTyped[*v1.ConfigMapList](cs.CoreV1().ConfigMaps("")),
```

Comment:
Have you also considered CiliumNodeConfig?

Reply:
Yes, the idea is to support multiple ConfigMaps including the LocalNodeConfig, I didn't have time today to rework this implementation but the goal is to refactor using the TransformMany and flatten the Map and the key would be something like (Key, Source) -> Value, and some logic to define the priority of each source, where Key is the key from the configmaps, the source is the configmap name, and Value is the key's values. 

So when we query this, we'll return the value from the ConfigMap with the highest priority.


I would prefer to rework this and have the logic for adding other ConfigMaps in this PR but adding the CiliumLocalNodeConfig to be in a later PR, I can still take care of it but I think having this piece in for now could be useful. Also having that PR as an example of how to add other ConfigMaps could facilitate later additions.

Reply:
What is your take on (2)?

Reply:
That sounds like a feasible approach, I advocate for adding such a way to notify in case of config drift. I think we can still have the logic for computing the deltas in the daemon. Or do you think it would fit better in the dynamicconfig pkg?

Reply:
My thinking is that we don't necessarily need to compute the diff outside of the dynamicconfig pkg since it's already implementing logic to fetch the desired state. It would then just be a matter of diff with the realized state and setting some flag (or something similar) that then the daemon can fetch in its status report.

Reply:
Let's continue the discussion on Slack, and I'll follow up with another PR for this part.